src/transformers/data/data_collator.py

```diff
@@ -66,6 +67,55 @@ def default_data_collator(features: List[InputDataClass]) -> Dict[str, torch.Ten
     return batch
 
 
+@dataclass
+class DataCollatorWithPadding:
+    """
+    Data collator that will dynamically pad the inputs received.
+
+    Args:
+        tokenizer (:class:`~transformers.PreTrainedTokenizer` or :class:`~transformers.PreTrainedTokenizerFast`):
+            The tokenizer used for encoding the data.
+        padding (:obj:`bool`, :obj:`str` or :class:`~transformers.tokenization_utils_base.PaddingStrategy`, `optional`, defaults to :obj:`True`):
+            Select a strategy to pad the returned sequences (according to the model's padding side and padding
+            index) among:
+
+            * :obj:`True` or :obj:`'longest'`: Pad to the longest sequence in the batch (or no padding if only a
+              single sequence if provided).
+            * :obj:`'max_length'`: Pad to a maximum length specified with the argument :obj:`max_length` or to the
+              maximum acceptable input length for the model if that argument is not provided.
+            * :obj:`False` or :obj:`'do_not_pad'` (default): No padding (i.e., can output a batch with sequences of
+              different lengths).
+        max_length (:obj:`int`, `optional`):
+            Maximum length of the returned list and optionally padding length (see above).
+        pad_to_multiple_of (:obj:`int`, `optional`):
+            If set will pad the sequence to a multiple of the provided value.
+
+            This is especially useful to enable the use of Tensor Cores on NVIDIA hardware with compute capability
+            >= 7.5 (Volta).
+    """
+
+    tokenizer: Union[PreTrainedTokenizer, PreTrainedTokenizerFast]
+    padding: Union[bool, str, PaddingStrategy] = True
+    max_length: Optional[int] = None
+    pad_to_multiple_of: Optional[int] = None
+
+    def __call__(self, features):
+        batch = self.tokenizer.pad(
```

Comment:
As it is coded now `features` cannot be of a `dict` of `torch.Tensor` `self.tokenizer.pad(...)` would fail, right?
This is a bit different from some other DataCollators that expect `torch.Tensors` as an input. 

Do we want to go more towards an approach where we preprocess everything in "Python" using the `nlp` library and then use the data collators to transformers a dataset of Python lists to a dataset of torch Tensors?

Reply:
I disagree that data collators expect `torch.Tensor` as inputs: `default_data_tokenizer` accept both. IMO it's the `DataCollatorForLanguageModeling` and variants that are not great in the sense they don't accept both tensors and list of ints as inputs. On that note, this data collator should also work with tensors, I agree. IMO it's the pad method that should be fixed to accept tensors ultimately.

I think we want to support everything in terms of preprocessing, but `nlp` should be a premium citizen.

Reply:
Ok, no I agree with you - I also think we should refactor the other data collators to also accept Python list inputs. 

IMO, a nice workflow is:
1) Download a dataset with `nlp`, this will be in the format `Dict[str, List[int or str]]`. 
2) Then we can use the dataset.map() functionality to process the data to `input_ids`, `attention_mask`, etc....Since `nlp.map(...)` does not allow to return `torch.Tensors`, here we have to stay anyways in Python List format.
3) This dataset can then ideally already be used with the `default_data_tokenizer` because all tokenization methods were applied during the .map() operations. 

Here would be an example of this workflow: https://huggingface.co/patrickvonplaten/bert2bert-cnn_dailymail-fp16#training-script.  (Note that the tensors would not even have to be put into torch tensor format). 
What do you think @sgugger ?

Reply:
I agree, except step 3 should be `default_data_collator` only when you are padding to a fixed max length. With the shuffle of the training set, you need dynamic padding as provided by the data collator introduced in this PR :-)